### PR TITLE
Make GenServer documentation friendlier to beginners

### PR DESCRIPTION
**Pull request by batate**
Per the discussion on Elixir Core. I am creating a pull request for further discussion. 

1. Use arguments and variables to describe what the arguments to GenServer callbacks and result tuples do. 
2. Call out the purpose of `c:init/1` by having it be more than a passthrough for the initial GenServer state. 
3. Do less work in the function heads and add more intermediate variables to label important beginner concepts. 

-bt

### CODE REVIEW

File: lib/elixir/lib/gen_server.ex [hunk 1]
```diff
@@ -15,62 +15,79 @@ defmodule GenServer do
...
+  a reply from the server (and are therefore synchronous) while **cast**
+  messages do not.
+
+  Each call to `GenServer.call/3` results in a message
   that must be handled by the `c:handle_call/3` callback in the GenServer.
-  A `cast/2` message must be handled by `c:handle_cast/2`. There are 8 possible
-  callbacks to be implemented when you use a `GenServer`. The only required
-  callback is `c:init/1`.
+  A `cast/2` message must be handled by `c:handle_cast/2`. `GenServer`
+  supports 8 callbacks, but only  `c:init/1` is required.
```
Comment (josevalim) [inline lib/elixir/lib/gen_server.ex:82]:
```suggestion
  supports 8 callbacks, but only `c:init/1` is required.
```

File: lib/elixir/lib/gen_server.ex [hunk 2]
```diff
@@ -15,62 +15,79 @@ defmodule GenServer do
...
+        def init(elements) do
+          initial_state = String.split(elements, ",", trim: true)
+          {:ok, initial_state}
         end
 
         @impl true
-        def handle_call(:pop, _from, [head | tail]) do
-          {:reply, head, tail}
+        def handle_call(:pop, _from, state) do
+          [to_caller|new_state] = state
```
Comment (josevalim) [inline lib/elixir/lib/gen_server.ex:41]:
```suggestion
          [to_caller | new_state] = state
```

File: lib/elixir/lib/gen_server.ex [hunk 3]
```diff
@@ -15,62 +15,79 @@ defmodule GenServer do
...
+          {:reply, to_caller, new_state}
         end
 
         @impl true
         def handle_cast({:push, element}, state) do
-          {:noreply, [element | state]}
+          new_state = [element|state]
```
Comment (josevalim) [inline lib/elixir/lib/gen_server.ex:47]:
```suggestion
          new_state = [element | state]
```

File: lib/elixir/lib/gen_server.ex [hunk 4]
```diff
@@ -94,21 +111,25 @@ defmodule GenServer do
...
+        def init(elements) do
+          initial_state = String.split(elements, ",", trim: true)
+          {:ok, initial_state}
         end
 
         @impl true
-        def handle_call(:pop, _from, [head | tail]) do
-          {:reply, head, tail}
+        def handle_call(:pop, _from, state) do
+          [to_caller|new_state] = state
```
Comment (josevalim) [inline lib/elixir/lib/gen_server.ex:121]:
```suggestion
          [to_caller | new_state] = state
```

File: lib/elixir/lib/gen_server.ex [hunk 5]
```diff
@@ -94,21 +111,25 @@ defmodule GenServer do
...
+          {:reply, to_caller, new_state}
         end
 
         @impl true
         def handle_cast({:push, element}, state) do
-          {:noreply, [element | state]}
+          new_state = [element|state]
```
Comment (josevalim) [inline lib/elixir/lib/gen_server.ex:127]:
```suggestion
          new_state = [element | state]
```

**Comment by josevalim** [comment]
:green_heart: :blue_heart: :purple_heart: :yellow_heart: :heart:

**Review by siestakimberly — COMMENTED** [review]
Seeing this change as I am going through emails. Thanks for making documentation beginner friendly :)

**Pull request merged**